### PR TITLE
Update memoized_helpers.rb

### PR DESCRIPTION
**Pull request by ioquatix**
I would like to write specs like this:

```ruby
	context '#localhost?' do
		let(:url_string) {"https://localhost"}
		let(:hostname) {self.class.metadata.fetch(:description_args).first}
		subject {Async::HTTP::Endpoint.parse(url_string, hostname: hostname)}
		
		describe 'localhost' do
			it "should be localhost" do
				is_expected.to be_localhost
			end
		end
		
		describe 'hello.localhost' do
			it "should be localhost" do
				is_expected.to be_localhost
			end
		end
		
		describe 'localhost.' do
			it "should be localhost" do
				is_expected.to be_localhost
			end
		end
		
		describe 'hello.localhost.' do
			it "should be localhost" do
				is_expected.to be_localhost
			end
		end
	end
```

but right now it's very cumbersome to get the actual inner most description.

### CODE REVIEW

**Comment by ioquatix** [comment]
This PR is very simple, I just wanted to make a proposal, if you think it's acceptable, I will improve the documentation and tests.

**Comment by pirj** [comment]
There is a non-cumbersome way to achieve what you are looking for.

```ruby
# spec/spec_helper.rb
RSpec.shared_context 'docstring as description do
  let(:description) {self.class.metadata.fetch(:description_args).first}
end

RSpec.configure |config|
  config.include_context 'docstring as description'
end

# spec/lib/localhost_spec.rb
context '#localhost?' do
  let(:hostname) {description}
  let(:url_string) {"https://localhost"}
  # ...
end
```

RSpec is very flexible, and you can even:

```ruby
# spec/spec_helper.rb
RSpec.shared_context 'extract docstring' do
  let(metadata[:extract_docstring]) { self.class.metadata.fetch(:description_args).first }
end

RSpec.configure do |config|
  config.include_context 'extract docstring', :extract_docstring
end

# spec/lib/localhost_spec.rb
RSpec.describe '#localhost?', extract_docstring: :hostname do
  describe 'localhost' do
    it 'should be localhost' do
      expect(hostname).to be_localhost
    end
  end
end
```

However, there are other styles I hope you'll consider.

TLDR;

1. There is some redundancy in the follwing part:

```ruby
it "should be localhost" do
  is_expected.to be_localhost
end
```

`is_expected` is, in general:
> Designed to read nicely for one-liners.

There would be no difference in the `--format documentation` output if you remove the docstring:

```ruby
it { is_expected.to be_localhost }
```

> RSpec will give the examples a doc string that is auto-generated from the matcher used in the example.
> This is designed specifically to help avoid duplication in situations where the doc string and the matcher used in the example mirror each other exactly.

A bit of history:

> `subject` was contributed by Joe Ferris to support the one-liner syntax embraced by shoulda matchers:
>       it { is_expected.to validate_presence_of(:name) }

2. Even though `context` and `describe` are interchangeable, they have different semantics.

`describe`:
> Generally used when grouping examples by a thing you are describing (e.g. an object, class or method).

`context`:
> Generally used when grouping examples contextually (e.g. "with xyz", "when xyz" or "if xyz").

Following that it would make sense to change the spec to use them differently:

```ruby
describe '#localhost?' do
  context 'hello.localhost' do
```

3. Helper method

There is an idea I've picked up from Effective Testing with RSpec book to use helper methods instead of setting up in hooks.
I understand that there are no explicit hooks in the code, but `let(:hostname)` definition is doubtfully cumbersome.
With helper method the spec might look like:

```ruby
def parse(hostname)
  # `url_string` defined in a `let` is available in this context
  Async::HTTP::Endpoint.parse(url_string, hostname: hostname)
end

it 'treats hello.localhost as localhost' do
  expect(parse('hello.localhost')).to be_localhost
end
```

4. One expectation per example

According to [this issue](https://github.com/rubocop-hq/rubocop-rspec/issues/228):
> the intent is one 'business' expectation per spec, not literally

Also keeping in mind the recently'ish introduction of `aggregate_failures`, an option that would show all failed expectations in the example, not only the first only.


With all this in mind, would you consider the following as an option?

```ruby
describe '#localhost?' do
  let(:url_string) {'https://localhost'}

  def parse(hostname)
    Async::HTTP::Endpoint.parse(url_string, hostname: hostname)
  end

  specify do
    expect(parse('localhost')).to be_localhost
    expect(parse('hello.localhost')).to be_localhost
    expect(parse('localhost.')).to be_localhost
    expect(parse('hello.localhost.')).to be_localhost
  end
end
```

**Comment by ioquatix** [comment]
Hey, I'm still groking your reply, but everything you say is awesome, thanks for taking the time to write it all out, it looks great.

**Comment by ioquatix** [comment]
this is what I ended up with

```ruby
	describe '#localhost?' do
		let(:url_string) {'https://localhost'}

		def endpoint_for(hostname)
			Async::HTTP::Endpoint.parse(url_string, hostname: hostname)
		end

		specify do
			expect(endpoint_for('localhost')).to be_localhost
			expect(endpoint_for('hello.localhost')).to be_localhost
			expect(endpoint_for('localhost.')).to be_localhost
			expect(endpoint_for('hello.localhost.')).to be_localhost
		end
		
		specify do
			expect(endpoint_for('localhost.com')).to_not be_localhost
		end
	end
```

I'm actually totally fine with that.

So, I don't know if this PR still has value or not. Maybe not.

**Comment by ioquatix** [comment]
The alternative is this:

```ruby
	describe '#localhost?' do
		let(:url_string) {"https://localhost"}
		let(:hostname) {self.class.metadata.fetch(:description_args).first}
		subject {Async::HTTP::Endpoint.parse(url_string, hostname: hostname)}
		
		context 'localhost' do
			it { is_expected.to be_localhost }
		end
		
		context 'hello.localhost' do
			it { is_expected.to be_localhost }
		end
		
		context 'localhost.' do
			it { is_expected.to be_localhost }
		end
		
		context 'hello.localhost.' do
			it { is_expected.to be_localhost }
		end
		
		context 'localhost.com' do
			it { is_expected.to_not be_localhost }
		end
	end
```

The output is a bit different:

```
  #localhost?
    localhost
      should be localhost
    hello.localhost
      should be localhost
    localhost.
      should be localhost
    hello.localhost.
      should be localhost
    localhost.com
      should not be localhost
  #localhost?
    should be localhost
    should not be localhost
```

I think the more explicit option is a bit nicer since you can see at a glance what is being tested.

**Comment by ioquatix** [comment]
I used your first suggestion:

```ruby
RSpec.shared_context 'docstring as description' do
	let(:description) {self.class.metadata.fetch(:description_args).first}
end

RSpec.configure do |config|
	# Enable flags like --only-failures and --next-failure
	config.example_status_persistence_file_path = ".rspec_status"

	config.include_context 'docstring as description'

	config.expect_with :rspec do |c|
		c.syntax = :expect
	end
end
```

And the spec:


```ruby
	describe '#localhost?' do
		subject {Async::HTTP::Endpoint.parse("https://localhost", hostname: description)}
		
		context 'localhost' do
			it { is_expected.to be_localhost }
		end
		
		context 'hello.localhost' do
			it { is_expected.to be_localhost }
		end
		
		context 'localhost.' do
			it { is_expected.to be_localhost }
		end
		
		context 'hello.localhost.' do
			it { is_expected.to be_localhost }
		end
		
		context 'localhost.com' do
			it { is_expected.to_not be_localhost }
		end
	end
```

I think that's the right blend of implicit vs explicit.

So, I am happy with that. I think there could be more discussion about this PR. I personally think it would be a useful addition, but I can see 1/ how I can do it without this PR which is awesome and 2/ there might be other approaches which are better.

**Comment by pirj** [comment]
Happy to hear that it helped, @ioquatix.
I admire your hard work on Async ecosystem.

**Comment by ioquatix** [comment]
@pirj what do you think of Rspec `its`?

e.g..

```ruby
	describe '#hostname' do
		describe Async::HTTP::Endpoint.parse("https://127.0.0.1:9292") do
			its(:hostname) {should == '127.0.0.1'}
		end
		
		describe Async::HTTP::Endpoint.parse("https://127.0.0.1:9292", hostname: 'localhost') do
			its(:hostname) {should == 'localhost'}
			it {is_expected.to_not be_localhost}
		end
	end
```

**Comment by JonRowe** [comment]
Thanks for taking the time to contribute this @ioquatix but this isn't something I'm going to merge because it's oddly specific and we don't really want to encourage accessing the docstrings in this fashion. They are there to describe your tests not to be part of them.

This kind of behaviour is also why shared examples exists, I would have rewritten this as:
```ruby
describe '#localhost?' do
  shared_examples_for "when host is" do |hostname|
    let(:url_string) { "https://localhost" }
    subject(:parse_result) { Async::HTTP::Endpoint.parse(url_string, hostname: hostname) }

    it "will return true" do
      is_expected.to be_localhost
    end
  end

  include_examples 'when host is', 'localhost'
  include_examples 'when host is', 'hello.localhost'
  include_examples 'when host is', 'localhost.'
  include_examples 'when host is', 'hello.localhost.'
end
```

**Comment by ioquatix** [comment]
Thanks. I’ve used shared examples like this before but they are a bit less flexible, however I’m happy using the globally shared context to provide description.

**Comment by JonRowe** [comment]
For the record, shared context and shared examples are the same thing, they are very flexible.

**Comment by ioquatix** [comment]
> For the record, shared context and shared examples are the same thing,

So, of course you can do anything you want in a shared example/context. I'll try to explain my POV more clearly. Not because I don't agree with you. But mostly just in case there is something worth discussing here further.

I think we all agree that DRY is a good principle.

It's why I like using descriptions to drive specs, especially when it's things like URLs or things that are being parsed. I think it's nice because not only is the description logically tied to the test, e.g. `describe "https://localhost" do ... `, it's clear what's breaking when a test fails.

When you write something like:

```ruby
describe '#localhost?' do
  shared_examples_for "when host is" do |hostname|
    let(:url_string) { "https://localhost" }
    subject(:parse_result) { Async::HTTP::Endpoint.parse(url_string, hostname: hostname) }

    it "will return true" do
      is_expected.to be_localhost
    end
  end

  include_examples 'when host is', 'localhost'
  include_examples 'when host is', 'hello.localhost'
  include_examples 'when host is', 'localhost.'
  include_examples 'when host is', 'hello.localhost.'
end
```

You: 1/ move the logic of the test away from what you are testing, 2/ loose meaningful description of what you are testing, 3/ make it hard to add specific expectations to individual cases.

Here is the output of what you suggested:

```
  #localhost?
    will return true
    will return true
    will return true
    will return true
```

Here is the output of what I ended up with:

```
  #localhost?
    http://localhost
      should be localhost
    http://hello.localhost
      should be localhost
    http://localhost.
      should be localhost
    http://hello.localhost.
      should be localhost
    http://localhost.com
      should not be localhost
```

For reference, here is the implementation:

```
	describe '#localhost?' do
		subject {Async::HTTP::Endpoint.parse(description)}
		
		context 'http://localhost' do
			it { is_expected.to be_localhost }
		end
		
		context 'http://hello.localhost' do
			it { is_expected.to be_localhost }
		end
		
		context 'http://localhost.' do
			it { is_expected.to be_localhost }
		end
		
		context 'http://hello.localhost.' do
			it { is_expected.to be_localhost }
		end
		
		context 'http://localhost.com' do
			it { is_expected.to_not be_localhost }
		end
	end
```

Along with the description shared context described by @pirj.

**Comment by JonRowe** [comment]
Yes well the `will return true` references `#localhost?` and was deliberate, as its considered that better doc strings will be explicit and not of the "maybe" style. e.g. `#localhost? when host is 'n' will return true` is a better output, but I'll admit I didn't expect `include_examples` to omit the context description however. A slight tweak to my shared example:

```
shared_examples_for "when host is" do |hostname|
  let(:url_string) { "https://localhost" }
  subject(:parse_result) { Struct.new(:localhost?).new(true) }

  it "will return true when hostname is #{hostname}" do
    is_expected.to be_localhost
  end
end
```

produces:

```
#localhost?
  will return true when hostname is localhost
  will return true when hostname is hello.localhost
  will return true when hostname is localhost.
  will return true when hostname is hello.localhost.
```
